hail/python/hailtop/aiotools/local_fs.py

```diff
@@ -110,6 +110,84 @@ def __str__(self) -> str:
         return 'file:' + self._path
 
 
+class TruncatedReadableBinaryIO(BinaryIO):
+    def __init__(self, bio: BinaryIO, limit: int):
+        self.bio = bio
+        self.n = 0
+        self.limit = limit
+
+    def write(self, s) -> int:  # pylint: disable=unused-argument
+        raise NotImplementedError
+
+    def __enter__(self) -> 'BinaryIO':
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.close()
+
+    @property
+    def mode(self) -> str:
+        return self.bio.mode
+
+    @property
+    def name(self) -> str:
+        return self.bio.name
+
+    def close(self) -> None:
+        return self.bio.close()
+
+    @property
+    def closed(self) -> bool:
+        return self.bio.closed
+
+    def fileno(self) -> int:
+        return self.bio.fileno()
+
+    def flush(self) -> None:
+        raise NotImplementedError
+
+    def isatty(self) -> bool:
+        return self.bio.isatty()
+
+    def read(self, n: int = -1):
+        assert self.n <= self.limit
+
+        if n == -1:
+            n = self.limit - self.n
+        else:
+            n = min(self.limit - self.n, n)
+        b = self.bio.read(self.limit - self.n)
```

Comment:
shouldn't this be `b = self.bio.read(n)`?

Reply:
yes